src/axom/sidre/core/ListCollection.hpp

```diff
@@ -13,9 +13,7 @@
  *          This is an implementation of ItemCollection to hold a
  *          collection of items of a fixed type. This implementation
  *          is intended to hold items that may have no name. If they do
- *          have names, those names are ignored. To satisfy the parent
- *          class interface, methods to access items by name are provided
- *          but they return null or invalid return values.
+ *          have names, those names are ignored.
```

Comment:
@nselliott -- When you get a chance, could you please check if `ListCollection` is still necessary, or if we should favor the new `IndexedCollection`.

The main difference is that `ListCollection` maintains a list of the used indexes, which provides fast O(1) insertions, but O(n) removals.  This is stress testing in `sidre_collections` tests that adds and removes 1000s of items and `ListCollection` is noticeably slower there.

In contrast, `IndexedCollection` has more complex logic to support adding items at a user-supplied index, which was necessary for creating `Buffer`s in the DataStore class.

Reply:
It looks like `IndexedCollection` satisfies the use case motivations for `ListCollection`, allowing a `Group` to hold a collection of unnamed objects, and iteration that preserves the order of insertion, if the items are inserted without user-specified indices.

What is the performance cost of insertion in `IndexedCollection` compared to `ListCollection`? O(1) is nice to have, but if we're not adding too much to that, it seems a good idea to eliminate one of the classes to avoid redundant functionality.  I think leaving it as a subsequent task and not adding it to this PR would be the way to go, though.

Reply:
Thanks @nselliott -- I agree about swapping it out in a follow-up PR.

Insertion in an `IndexedCollection` is amortized O(1) -- if the index is in range, it will pop the `free_id` stack, otherwise, it grows the size of the array (and fills with `nullptr` to match the index.  The only slightly complicated part is that indices in the `free_ids` list might be occupied, so we need to check `hasItem`  before using them.